src/Microsoft.Azure.WebJobs.Host/Scale/InMemoryScaleMetricsRepository.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    public class InMemoryScaleMetricsRepository : IScaleMetricsRepository
+    {
+        private readonly IDictionary<IScaleMonitor, IList<ScaleMetrics>> _monitorMetrics;
+        private readonly TimeSpan _scaleMetricsMaxAge;
+        private readonly ILogger _logger;
+
+        public InMemoryScaleMetricsRepository(ILoggerFactory loggerFactory)
+        {
+            _monitorMetrics = new Dictionary<IScaleMonitor, IList<ScaleMetrics>>();
+            _scaleMetricsMaxAge = TimeSpan.FromMinutes(2);
+            _logger = loggerFactory.CreateLogger<InMemoryScaleMetricsRepository>();
+        }
+
+        public Task<IDictionary<IScaleMonitor, IList<ScaleMetrics>>> ReadMetricsAsync(IEnumerable<IScaleMonitor> monitors)
+        {
+            lock (_monitorMetrics)
+            {
+                IDictionary<IScaleMonitor, IList<ScaleMetrics>> result = new Dictionary<IScaleMonitor, IList<ScaleMetrics>>();
+
+                // Clean old metrics
+                DateTime threshold = DateTime.Now - _scaleMetricsMaxAge;
+                foreach (IList<ScaleMetrics> metrics in _monitorMetrics.Values)
+                {
+                    foreach (ScaleMetrics metric in metrics)
+                    {
+                        if (metric.Timestamp < threshold)
+                        {
+                            metrics.Remove(metric);
+                        }
+                    }
+                }
+
+                foreach (IScaleMonitor monitor in monitors)
+                {
+                    result[monitor] = new List<ScaleMetrics>();
+                    Type metricsType = GetMonitorScaleMetricsTypeOrNull(monitor);
+                    if (metricsType != null)
+                    {
+                        result[monitor] = _monitorMetrics.ContainsKey(monitor) ? _monitorMetrics[monitor] : result[monitor];
+                    }
+                }
+                return Task.FromResult(result);
+            }
+        }
+
+        public Task WriteMetricsAsync(IDictionary<IScaleMonitor, ScaleMetrics> monitorMetrics)
+        {
+            lock (_monitorMetrics)
```

Comment:
Only writes to the same IScaleMonitor should be locked right? (Instead of locking the entire metrics cache) We should allow concurrent writes to different keys in the dictionary -- perhaps a `ConcurrentDictionary` is a better data structure for this use case.

Reply:
`InMemoryScaleMatricsRepository` is moving to `ScaleController.Common`